src/optimal_control_model-getters.jl

```diff
@@ -122,66 +143,116 @@ function nlp_constraints!(ocp::OptimalControlModel)
         val = zeros(ctNumber, dim)
         j = 1
         for i ∈ 1:length(ξf)
+            li = ξs[i]
             vali = ξf[i](t, u, v)
-            li = length(vali)
             val[j:(j + li - 1)] .= vali # .= also allows scalar value for vali
             j = j + li
         end
         return val
     end
 
+    function ξ!(val, t, u, v) # nonlinear control constraints (in place)
+        j = 1
+        for i ∈ 1:length(ξf)
+            li = ξs[i]
+            ξf[i](val[j:(j + li - 1)], t, u, v)
+            j = j + li
+        end
+        return nothing
+    end
+
     function η(t, x, v) # nonlinear state constraints
         dim = length(ηl)
         val = zeros(ctNumber, dim)
         j = 1
         for i ∈ 1:length(ηf)
+            li = ηs[i]
             vali = ηf[i](t, x, v)
-            li = length(vali)
             val[j:(j + li - 1)] .= vali # .= also allows scalar value for vali
             j = j + li
         end
         return val
     end
 
+    function η!(val, t, x, v) # nonlinear state constraints (in place)
+        j = 1
+        for i ∈ 1:length(ηf)
+            li = ηs[i]
+            ηf[i](val[j:(j + li - 1)], t, x, v)
+            j = j + li
+        end
+        return nothing
+    end
+
     function ψ(t, x, u, v) # nonlinear mixed constraints
         dim = length(ψl)
         val = zeros(ctNumber, dim)
         j = 1
         for i ∈ 1:length(ψf)
+            li = ψs[i]
             vali = ψf[i](t, x, u, v)
-            li = length(vali)
             val[j:(j + li - 1)] .= vali # .= also allows scalar value for vali
             j = j + li
         end
         return val
     end
 
+    function ψ!(val, t, x, u, v) # nonlinear mixed constraints (in place)
+        j = 1
+        for i ∈ 1:length(ψf)
+            li = ψs[i]
+            ψf[i](val[j:(j + li - 1)], t, x, u, v)
```

Comment:
I thing that `val[j:(j + li - 1)]` will allocate a new vector.
You probably need `view(val, j:(j + li - 1))`.

Reply:
@amontoison this is subtle... any doc / reference on this?

Reply:
NB. The syntax with `@view` is supposed to generate the same code (and allows `begin` / `end` indexing):
```julia
julia> g4() = begin
       x = zeros(Float64, 10)
       for i ∈ 1:1000 f!(@view x[3:4]) end
       println("x =", x)
       end
g4 (generic function with 1 method)

julia> @allocated g4()
x =[0.0, 0.0, 7.0, 8.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
5728
```

Reply:
```julia
julia> x = zeros(4)
4-element Vector{Float64}:
 0.0
 0.0
 0.0
 0.0

julia> y = ones(4)
4-element Vector{Float64}:
 1.0
 1.0
 1.0
 1.0

julia> @allocated x[1:2] = y[3:4]
80

julia> @allocated x[1:2] = y[3:4]
80

julia> @allocated x[1:2] .= y[3:4]
192

julia> @allocated x[1:2] .= y[3:4]
192

julia> @allocated x[1:2] = @views y[3:4]
48

julia> @allocated x[1:2] = @views y[3:4]
48
```

Reply:
Don't test things at a global scope @jbcaillau.
Or wrap it in a function of a `let ... end` block.
https://discourse.julialang.org/t/using-allocated-to-track-memory-allocations/4617

Reply:
```julia
ulia> let 
       x = zeros(4)
       y = ones(4)
       println(@allocated x[1:2] = y[3:4])
       x
       end
80
4-element Vector{Float64}:
 1.0
 1.0
 0.0
 0.0

julia> let 
       x = zeros(4)
       y = ones(4)
       println(@allocated x[1:2] .= y[3:4])
       x
       end
80
4-element Vector{Float64}:
 1.0
 1.0
 0.0
 0.0

julia> let 
       x = zeros(4)
       y = ones(4)
       println(@allocated x[1:2] = @views y[3:4])
       x
       end
0
4-element Vector{Float64}:
 1.0
 1.0
 0.0
 0.0
```

Reply:
Yes, they do a copy in both cases

Reply:
online https://univ-cotedazur.zoom.us/my/caillau

Reply:
@PierreMartinon check above: slices in the rhs of an affectation `x[...] = y[...]` are copied unless a view is used, `x[...] = view(y, ...)`

Reply:
Yes I know... Julia compiler right now is not great at optimizations. See also: https://discourse.julialang.org/t/common-allocation-mistakes/66127

Reply:
https://m3g.github.io/JuliaNotes.jl/stable/memory